[BUG] Bad document JSON data leads to a dangling promise returned by `pull_session`

Suppose I have created a document with a non-existing model. When the `pull_session` is called, I see these messages in the JS console:
```
[bokeh] Got an event for unknown model  {id: "1053", type: "LinearColorMapper"}
Document.apply_json_patch @ bokeh.js?v=1.3.4:8503
ClientSession._handle_patch @ bokeh.js?v=1.3.4:528
ClientSession.handle @ bokeh.js?v=1.3.4:477
(anonymous) @ bokeh.js?v=1.3.4:317
Promise.then (async)
ClientConnection._repull_session_doc @ bokeh.js?v=1.3.4:299
ClientConnection._awaiting_ack_handler @ bokeh.js?v=1.3.4:405
_current_handler @ bokeh.js?v=1.3.4:347
ClientConnection._on_message @ bokeh.js?v=1.3.4:365
_this.socket.onmessage @ bokeh.js?v=1.3.4:217
```
```
21:37:36.553 backend.js:6 Error: event model wasn't known
    at Document.apply_json_patch (bokeh.js?v=1.3.4:8504)
    at ClientSession._handle_patch (bokeh.js?v=1.3.4:528)
    at ClientSession.handle (bokeh.js?v=1.3.4:477)
    at bokeh.js?v=1.3.4:317
r @ backend.js:6
(anonymous) @ bokeh.js?v=1.3.4:338
Promise.catch (async)
ClientConnection._repull_session_doc @ bokeh.js?v=1.3.4:336
ClientConnection._awaiting_ack_handler @ bokeh.js?v=1.3.4:405
_current_handler @ bokeh.js?v=1.3.4:347
ClientConnection._on_message @ bokeh.js?v=1.3.4:365
_this.socket.onmessage @ bokeh.js?v=1.3.4:217
```
```
21:37:36.553 bokeh.js?v=1.3.4:339 [bokeh] Failed to repull session Error: event model wasn't known
(anonymous) @ bokeh.js?v=1.3.4:339
Promise.catch (async)
ClientConnection._repull_session_doc @ bokeh.js?v=1.3.4:336
ClientConnection._awaiting_ack_handler @ bokeh.js?v=1.3.4:405
_current_handler @ bokeh.js?v=1.3.4:347
ClientConnection._on_message @ bokeh.js?v=1.3.4:365
_this.socket.onmessage @ bokeh.js?v=1.3.4:217
21:40:08.999 
```

It's pretty clear what the culprit is in my particular case.
However, the issue in Bokeh is that in this case the promise returned by `pull_session` is never resolved or rejected. And it makes it seem like `add_document_from_session` is still busy.

The root cause, I believe, is that `ClientConnection._repull_session_doc` does not call any callback if it fails - it just logs the error and that's it.
The constructor of `ClientConnection` doesn't even have any such callback for errors not related to a closed connection.
It's not really related to the issue itself, but there's also some strange code that:
1. Uses both the 2-argument form of `Promise.then` and the `Promise.catch`, which makes no sense
2. Makes it possible to call `reject` after `resolve` has already been called

With that being said, this is probably a very low priority issue since only some specific errors can trigger it, and even then it's should be clear to the developer what to do. It's just that the custom error handling code that someone's app may rely on will never be called.

--- a/bokehjs/src/lib/client/connection.ts
+++ b/bokehjs/src/lib/client/connection.ts
@@ -108,38 +108,25 @@ export class ClientConnection {
     message.send(this.socket)
   }
 
-  send_with_reply(message: Message): Promise<Message> {
+  async send_with_reply(message: Message): Promise<Message> {
     const promise = new Promise<Message>((resolve, reject) => {
       this._pending_replies[message.msgid()] = [resolve, reject]
       this.send(message)
     })
 
-    return promise.then(
-      (message) => {
-        if (message.msgtype() === "ERROR")
-          throw new Error(`Error reply ${message.content.text}`)
-        else
-          return message
-      },
-      (error) => {
-        throw error
-      },
-    )
+    const message = await promise
+    if (message.msgtype() === "ERROR")
+      throw new Error(`Error reply ${message.content.text}`)
+    else
+      return message
   }
 
-  protected _pull_doc_json(): Promise<DocJson> {
+  protected async _pull_doc_json(): Promise<DocJson> {
     const message = Message.create("PULL-DOC-REQ", {})
-    const promise = this.send_with_reply(message)
-    return promise.then(
-      (reply) => {
-        if (!('doc' in reply.content))
-          throw new Error("No 'doc' field in PULL-DOC-REPLY")
-        return reply.content.doc
-      },
-      (error) => {
-        throw error
-      },
-    )
+    const reply = await this.send_with_reply(message)
+    if (!("doc" in reply.content))
+      throw new Error("No 'doc' field in PULL-DOC-REPLY")
+    return reply.content.doc
   }
 
   protected _repull_session_doc(): void {

--- a/bokehjs/src/lib/client/connection.ts
+++ b/bokehjs/src/lib/client/connection.ts
@@ -10,6 +10,9 @@ export const DEFAULT_SESSION_ID = "default"
 let _connection_count: number = 0
 
 export type Rejecter = (error: Error | string) => void
+export type SessionResolver = (s: ClientSession) => void
+export type MessageResolver = (m: Message) => void
+export type PendingReply = {resolve: MessageResolver, reject: Rejecter}
 
 export class ClientConnection {
 
@@ -21,28 +24,25 @@ export class ClientConnection {
   closed_permanently: boolean = false
 
   protected _current_handler: ((message: Message) => void) | null = null
-  protected _pending_ack: [(connection: ClientConnection) => void, Rejecter] | null = null // null or [resolve,reject]
-  protected _pending_replies: {[key: string]: [(message: Message) => void, Rejecter]} = {} // map reqid to [resolve,reject]
+  protected _pending_replies: Map<string, PendingReply> = new Map()
   protected _pending_messages: Message[] = []
   protected readonly _receiver: Receiver = new Receiver()
 
   constructor(readonly url: string = DEFAULT_SERVER_WEBSOCKET_URL,
               readonly id: string = DEFAULT_SESSION_ID,
-              readonly args_string: string | null = null,
-              protected _on_have_session_hook: ((session: ClientSession) => void) | null = null,
-              protected _on_closed_permanently_hook: (() => void) | null = null) {
-
+              readonly args_string: string | null = null) {
     logger.debug(`Creating websocket ${this._number} to '${this.url}' session '${this.id}'`)
   }
 
-  connect(): Promise<ClientConnection> {
+  async connect(): Promise<ClientSession> {
     if (this.closed_permanently)
-      return Promise.reject(new Error("Cannot connect() a closed ClientConnection"))
+      throw new Error("Cannot connect() a closed ClientConnection")
     if (this.socket != null)
-      return Promise.reject(new Error("Already connected"))
+      throw new Error("Already connected")
 
-    this._pending_replies = {}
     this._current_handler = null
+    this._pending_replies.clear()
+    this._pending_messages = []
 
     try {
       let versioned_url = `${this.url}?bokeh-session-id=${this.id}`
@@ -57,13 +57,13 @@ export class ClientConnection {
         this.socket!.binaryType = "arraybuffer"
         this.socket!.onopen = () => this._on_open(resolve, reject)
         this.socket!.onmessage = (event) => this._on_message(event)
-        this.socket!.onclose = (event) => this._on_close(event)
+        this.socket!.onclose = (event) => this._on_close(event, reject)
         this.socket!.onerror = () => this._on_error(reject)
       })
     } catch (error) {
       logger.error(`websocket creation failed to url: ${this.url}`)
       logger.error(` - ${error}`)
-      return Promise.reject(error)
+      throw error
     }
   }
 
@@ -74,10 +74,6 @@ export class ClientConnection {
       if (this.socket != null)
         this.socket.close(1000, `close method called on ClientConnection ${this._number}`)
       this.session!._connection_closed()
-      if (this._on_closed_permanently_hook != null) {
-        this._on_closed_permanently_hook()
-        this._on_closed_permanently_hook = null
-      }
     }
   }
 
@@ -109,16 +105,15 @@ export class ClientConnection {
   }
 
   async send_with_reply(message: Message): Promise<Message> {
-    const promise = new Promise<Message>((resolve, reject) => {
-      this._pending_replies[message.msgid()] = [resolve, reject]
+    const reply = await new Promise<Message>((resolve, reject) => {
+      this._pending_replies.set(message.msgid(), {resolve, reject})
       this.send(message)
     })
 
-    const message = await promise
-    if (message.msgtype() === "ERROR")
-      throw new Error(`Error reply ${message.content.text}`)
+    if (reply.msgtype() === "ERROR")
+      throw new Error(`Error reply ${reply.content.text}`)
     else
-      return message
+      return reply
   }
 
   protected async _pull_doc_json(): Promise<DocJson> {
@@ -129,65 +124,53 @@ export class ClientConnection {
     return reply.content.doc
   }
 
-  protected _repull_session_doc(): void {
-    if (this.session == null)
-      logger.debug("Pulling session for first time")
-    else
-      logger.debug("Repulling session")
-    this._pull_doc_json().then(
-      (doc_json) => {
-        if (this.session == null) {
-          if (this.closed_permanently)
-            logger.debug("Got new document after connection was already closed")
-          else {
-            const document = Document.from_json(doc_json)
-
-            // Constructing models changes some of their attributes, we deal with that
-            // here. This happens when models set attributes during construction
-            // or initialization.
-            const patch = Document._compute_patch_since_json(doc_json, document)
-            if (patch.events.length > 0) {
-              logger.debug(`Sending ${patch.events.length} changes from model construction back to server`)
-              const patch_message = Message.create('PATCH-DOC', {}, patch)
-              this.send(patch_message)
-            }
-
-            this.session = new ClientSession(this, document, this.id)
-
-            for (const msg of this._pending_messages) {
-              this.session.handle(msg)
-            }
-            this._pending_messages = []
-
-            logger.debug("Created a new session from new pulled doc")
-            if (this._on_have_session_hook != null) {
-              this._on_have_session_hook(this.session)
-              this._on_have_session_hook = null
-            }
-          }
+  protected async _repull_session_doc(resolve: SessionResolver, reject: Rejecter): Promise<void> {
+    logger.debug(this.session ? "Repulling session" : "Pulling session for first time")
+    try {
+      const doc_json = await this._pull_doc_json()
+      if (this.session == null) {
+        if (this.closed_permanently) {
+          logger.debug("Got new document after connection was already closed")
+          reject(new Error("The connection has been closed"))
         } else {
-          this.session.document.replace_with_json(doc_json)
-          logger.debug("Updated existing session with new pulled doc")
+          const document = Document.from_json(doc_json)
+
+          // Constructing models changes some of their attributes, we deal with that
+          // here. This happens when models set attributes during construction
+          // or initialization.
+          const patch = Document._compute_patch_since_json(doc_json, document)
+          if (patch.events.length > 0) {
+            logger.debug(`Sending ${patch.events.length} changes from model construction back to server`)
+            const patch_message = Message.create('PATCH-DOC', {}, patch)
+            this.send(patch_message)
+          }
+
+          this.session = new ClientSession(this, document, this.id)
+
+          for (const msg of this._pending_messages) {
+            this.session.handle(msg)
+          }
+          this._pending_messages = []
+
+          logger.debug("Created a new session from new pulled doc")
+          resolve(this.session)
         }
-      },
-      (error) => {
-        // handling the error here is useless because we wouldn't
-        // get errors from the resolve handler above, so see
-        // the catch below instead
-        throw error
-      },
-    ).catch((error) => {
-      if (console.trace != null)
-        console.trace(error)
+      } else {
+        this.session.document.replace_with_json(doc_json)
+        logger.debug("Updated existing session with new pulled doc")
+        // Since the session already exists, we don't need to call `resolve` again.
+      }
+    } catch (error) {
+      console.trace?.(error)
       logger.error(`Failed to repull session ${error}`)
-    })
+      reject(error)
+    }
   }
 
-  protected _on_open(resolve: (connection: ClientConnection) => void, reject: Rejecter): void {
+  protected _on_open(resolve: SessionResolver, reject: Rejecter): void {
     logger.info(`Websocket connection ${this._number} is now open`)
-    this._pending_ack = [resolve, reject]
     this._current_handler = (message: Message) => {
-      this._awaiting_ack_handler(message)
+      this._awaiting_ack_handler(message, resolve, reject)
     }
   }
 
@@ -201,47 +184,34 @@ export class ClientConnection {
       this._close_bad_protocol(e.toString())
     }
 
-    if (this._receiver.message == null)
-      return
-
     const msg = this._receiver.message
+    if (msg != null) {
+      const problem = msg.problem()
+      if (problem != null)
+        this._close_bad_protocol(problem)
 
-    const problem = msg.problem()
-    if (problem != null)
-      this._close_bad_protocol(problem)
-
-    this._current_handler!(msg)
+      this._current_handler!(msg)
+    }
   }
 
-  protected _on_close(event: CloseEvent): void {
+  protected _on_close(event: CloseEvent, reject: Rejecter): void {
     logger.info(`Lost websocket ${this._number} connection, ${event.code} (${event.reason})`)
     this.socket = null
 
-    if (this._pending_ack != null) {
-      this._pending_ack[1](new Error(`Lost websocket connection, ${event.code} (${event.reason})`))
-      this._pending_ack = null
-    }
+    this._pending_replies.forEach((pr) => pr.reject("Disconnected"))
+    this._pending_replies.clear()
 
-    const pop_pending = () => {
-      for (const reqid in this._pending_replies) {
-        const promise_funcs = this._pending_replies[reqid]
-        delete this._pending_replies[reqid]
-        return promise_funcs
-      }
-      return null
-    }
-    let promise_funcs = pop_pending()
-    while (promise_funcs != null) {
-      promise_funcs[1]("Disconnected")
-      promise_funcs = pop_pending()
-    }
     if (!this.closed_permanently)
       this._schedule_reconnect(2000)
+
+    reject(new Error(`Lost websocket connection, ${event.code} (${event.reason})`))
   }
 
   protected _on_error(reject: Rejecter): void {
     logger.debug(`Websocket error on socket ${this._number}`)
-    reject(new Error("Could not open websocket"))
+    const msg = "Could not open websocket"
+    logger.error(`Failed to connect to Bokeh server: ${msg}`)
+    reject(new Error(msg))
   }
 
   protected _close_bad_protocol(detail: string): void {
@@ -250,26 +220,22 @@ export class ClientConnection {
       this.socket.close(1002, detail) // 1002 = protocol error
   }
 
-  protected _awaiting_ack_handler(message: Message): void {
+  protected _awaiting_ack_handler(message: Message, resolve: SessionResolver, reject: Rejecter): void {
     if (message.msgtype() === "ACK") {
       this._current_handler = (message: Message) => this._steady_state_handler(message)
 
       // Reload any sessions
-      this._repull_session_doc()
-
-      if (this._pending_ack != null) {
-        this._pending_ack[0](this)
-        this._pending_ack = null
-      }
+      this._repull_session_doc(resolve, reject)
     } else
       this._close_bad_protocol("First message was not an ACK")
   }
 
   protected _steady_state_handler(message: Message): void {
-    if (message.reqid() in this._pending_replies) {
-      const promise_funcs = this._pending_replies[message.reqid()]
-      delete this._pending_replies[message.reqid()]
-      promise_funcs[0](message)
+    const reqid = message.reqid()
+    const pr = this._pending_replies.get(reqid)
+    if (pr) {
+      this._pending_replies.delete(reqid)
+      pr.resolve(message)
     } else if (this.session) {
       this.session.handle(message)
     } else {
@@ -278,35 +244,7 @@ export class ClientConnection {
   }
 }
 
-// Returns a promise of a ClientSession
-// The returned promise has a close() method in case you want to close before
-// getting a session; session.close() works too once you have a session.
 export function pull_session(url?: string, session_id?: string, args_string?: string): Promise<ClientSession> {
-  const promise = new Promise<ClientSession>((resolve, reject) => {
-    const connection = new ClientConnection(
-      url, session_id, args_string,
-      (session) => {
-        try {
-          resolve(session)
-        } catch (error) {
-          logger.error(`Promise handler threw an error, closing session ${error}`)
-          session.close()
-          throw error
-        }
-      },
-      () => {
-        // we rely on reject() as a no-op if we already resolved
-        reject(new Error("Connection was closed before we successfully pulled a session"))
-      },
-    )
-    connection.connect().then(
-      (_) => undefined,
-      (error) => {
-        logger.error(`Failed to connect to Bokeh server ${error}`)
-        throw error
-      },
-    )
-  })
-
-  return promise
+  const connection = new ClientConnection(url, session_id, args_string)
+  return connection.connect()
 }

--- a/bokehjs/src/lib/client/session.ts
+++ b/bokehjs/src/lib/client/session.ts
@@ -35,10 +35,10 @@ export class ClientSession {
   // Sends a request to the server for info about the server, such as its Bokeh
   // version. Returns a promise, the value of the promise is a free-form dictionary
   // of server details.
-  request_server_info(): Promise<unknown> {
+  async request_server_info(): Promise<unknown> {
     const message = Message.create('SERVER-INFO-REQ', {})
-    const promise = this._connection.send_with_reply(message)
-    return promise.then((reply) => reply.content)
+    const reply = await this._connection.send_with_reply(message)
+    return reply.content
   }
 
   // Sends some request to the server (no guarantee about which one) and returns
@@ -50,8 +50,8 @@ export class ClientSession {
   // results of that processing without a race condition. (This assumes the
   // server processes events in sequence, which it mostly has to semantically,
   // since reordering events might change the final state.)
-  force_roundtrip(): Promise<void> {
-    return this.request_server_info().then((_) => undefined)
+  async force_roundtrip(): Promise<void> {
+    await this.request_server_info()
   }
 
   protected _document_changed(event: DocumentChangedEvent): void {

